package net.blockomorph.utils;

import net.blockomorph.BlockomorphServer;
import net.minecraft.nbt.CompoundTag;
import net.minecraft.nbt.NbtIo;

import java.io.File;
import java.util.HashMap;
import java.util.Map;

public class SavedBlockManager {
	private final File gameDir;
	private boolean init;
	private final HashMap<String, SavedBlock> blocks = new HashMap<>();

	public SavedBlockManager(File gameDir) {
		this.gameDir = new File(gameDir, "saved_blocks.bmdat");
	}

	public void load() {
		if (!init)
			try {
				CompoundTag tag = NbtIo.read(this.gameDir);
				if (tag != null) {
					for (String key : tag.getAllKeys()) {
						this.blocks.put(key, SavedBlock.fromTag(tag.getCompound(key), key));
					}
				}
				init = true;
			} catch (Exception e) {
				BlockomorphServer.LOGGER.error("An error occurred while loading saved blocks", e);
			}
	}

	public HashMap<String, SavedBlock> get() {
		return this.blocks;
	}

	public void delete(String s) {
		if (this.blocks.containsKey(s)) {
			this.blocks.remove(s);
			this.write();
		}
	}

	public void add(SavedBlock b) {
		String s = b.getName();
		if (!this.blocks.containsKey(s)) {
			this.blocks.put(s, b);
			this.write();
		}
	}

	public void clear() {
		this.blocks.clear();
		this.write();
	}

	public void write() {
		try {
			CompoundTag tag = new CompoundTag();
			for (Map.Entry<String, SavedBlock> entry : this.blocks.entrySet()) {
				tag.put(entry.getKey(), entry.getValue().toNbt());
			}
			NbtIo.write(tag, this.gameDir);
		} catch (Exception e) {
			BlockomorphServer.LOGGER.error("An error occurred while saving favourite blocks", e);
		}
	}
}
